compiler/src/lnntoamm/Fn.ts

```diff
@@ -219,22 +247,43 @@ export default class Fn {
       }
       stmt.inline(amm);
     }
+    this.retTy.resetTemp();
     this.params.forEach((param) => param.unassign());
   }
 
-  resultTyFor(argTys: Type[], scope: Scope): Type | null {
-    let res: Type | null = null;
+  resultTyFor(
+    argTys: Type[],
+    expectResTy: Type,
+    scope: Scope,
+  ): [Type[], Type] | null {
+    let res: [Type[], Type] | null = null;
     try {
-      this.params.forEach((param, ii) =>
-        param.ty.tempConstrain(argTys[ii], scope),
-      );
-      res = this.retTy.instance();
-    } catch (_e) {
+      this.params.forEach((param, ii) => {
+        return param.ty.tempConstrain(argTys[ii], scope);
+      });
+      this.retTy.tempConstrain(expectResTy, scope);
+      res = [
+        this.params.map((param) => param.ty.instance({ interfaceOk: true })),
+        this.retTy.instance({ interfaceOk: true }),
+      ];
+    } catch (e) {
       // do nothing: the args aren't applicable to the params so
       // we return null (`res` is already `null`) and we need to
       // ensure the param tys have `resetTemp` called on them.
+      // However, if the Error message starts with `TODO`, then
+      // print the error since it's for debugging purposes
+      const msg = e.message as string;
+      if (msg.startsWith('TODO')) {
+        console.log(
+          'warning: came across TODO from Types.ts when getting result ty for a function:',
+        );
+        console.group();
```

Comment:
Should this be above the other `console.log` to make sure they don't get split from each other?

Reply:
I use the group to indicate that the next line is associated to the warning. for example:
```
warning: came across TODO from Types.ts when getting result ty for a function:
  TODO: re-tempConstrain Interface
```